### PR TITLE
test the parallel compiler

### PR DESCRIPTION
**Pull request by SparrowLii**
part of #101566 
This test the perf of parallel compiler with some optimizations

### CODE REVIEW

**Comment by SparrowLii** [comment]
In order to prevent the impact of noise on walltime, let's test again

File: compiler/rustc_data_structures/src/sync.rs [hunk 1]
```diff
@@ -407,54 +611,79 @@ impl<K: Eq + Hash, V: Eq, S: BuildHasher> HashMapExt<K, V> for HashMap<K, V, S>
 }
 
 #[derive(Debug)]
-pub struct Lock<T>(InnerLock<T>);
+pub struct Lock<T> {
+    single_thread: bool,
+    inner: RefCell<T>,
+    mt_inner: Option<Mutex<T>>,
+}
```
Comment (nnethercote) [inline compiler/rustc_data_structures/src/sync.rs:600]: I have strong reservations about this whole approach of having one synchronization mechanism for single-threaded and another for multi-threaded.

Putting that aside for a moment: is it possible to make this type an enum?
```
pub enum Lock<T> {
    Single(RefCell<T>),
    Multi(Mutex<T>),
}
```
The type would be smaller, simpler, and less error-prone.

Comment (SparrowLii) [reply compiler/rustc_data_structures/src/sync.rs:600]: I tried this method at the beginning, but the effect looks very unsatisfactory:
https://github.com/rust-lang/rust/pull/101566#issuecomment-1241293886
and it is not even as good as using `std::sync::Mutex` directly:
https://github.com/rust-lang/rust/pull/101566#issuecomment-1276331871

Comment (SparrowLii) [reply compiler/rustc_data_structures/src/sync.rs:600]: This seems to be caused by the optimization problem of `LockGuard`. If we can research a way to get better optimization(@Zoxc has a exprimental branch for it), I think it is feasible.

Comment (nnethercote) [reply compiler/rustc_data_structures/src/sync.rs:600]: What is the problem with `LockGuard`? #101566 is very long so it is hard for me to see.

File: compiler/rustc_data_structures/src/sync.rs [hunk 2]
```diff
@@ -54,7 +59,43 @@ pub use std::sync::atomic::Ordering::SeqCst;
...
+    use std::sync::atomic::AtomicU8;
+
+    const UNINITIALIZED: u8 = 0;
+    const INACTIVE: u8 = 1;
+    const ACTIVE: u8 = 2;
+
+    static MODE: AtomicU8 = AtomicU8::new(UNINITIALIZED);
+
+    #[inline]
+    pub fn active() -> bool {
```
Comment (nnethercote) [inline compiler/rustc_data_structures/src/sync.rs:77]: The names in this module are extremely generic:
```
mod mode    
static MODE    
const UNINITIALIZED/ACTIVE/INACTIVE 
fn active()
fn set()
```
They tell the reader almost no information. I suggest these more descriptive names:
```
mod threadedness
static THREADEDNESS
const UNINITIALIZED/SINGLE/MULTI
fn is_single_threaded() / is_multi_threaded()
fn set_threadedness()
```

Comment (SparrowLii) [reply compiler/rustc_data_structures/src/sync.rs:77]: This was decided in https://github.com/rust-lang/rust/pull/107586#discussion_r1122591336 and some comments below that.
This is to use `sync::active` to indicate that only the data structures in the `sync` module are involved, not the entire compiler (whether in parallel mode or not).

File: compiler/rustc_data_structures/src/sync.rs (hunk 1 above)
Comment (SparrowLii) [reply compiler/rustc_data_structures/src/sync.rs:600]: It's the comment https://github.com/rust-lang/rust/pull/101566#issuecomment-1407352906

Comment (SparrowLii) [reply compiler/rustc_data_structures/src/sync.rs:600]: From my local test, it is mainly the `drop` function of `LockGuard` that affects the optimization of llvm

File: compiler/rustc_data_structures/src/sync.rs (hunk 2 above)
Comment (nnethercote) [reply compiler/rustc_data_structures/src/sync.rs:77]: I disagree with that decision. If I see code like this:
```
if rustc_data_structures::sync::active() { ... }
```
it is unclear what `active` means. When I first looked at the code, I didn't understand it, and I had to go find the definition of `active` to understand it.

In comparison, this:
```
if rustc_data_structures::sync::is_multi_threaded() { ... }
```
is much clearer. Good names are important!

Comment (SparrowLii) [reply compiler/rustc_data_structures/src/sync.rs:77]: Yea makes sense. Let me make a change and add a comment in #107586

Comment (Zoxc) [reply compiler/rustc_data_structures/src/sync.rs:77]: They control thread safetly, not thread count so `is_multi_threaded` is a bit misleading.

Comment (Zoxc) [reply compiler/rustc_data_structures/src/sync.rs:77]: Maybe `dyn_thread_safe` or `is_dyn_thread_safe`?

Comment (nnethercote) [reply compiler/rustc_data_structures/src/sync.rs:77]: `Lock` also has a field called `single_thread`. Should it also be called `is_dyn_thread_safe`

Comment (SparrowLii) [reply compiler/rustc_data_structures/src/sync.rs:77]: I agree with `is_dyn_thread_safe`

File: compiler/rustc_interface/src/interface.rs [hunk 3]
```diff
@@ -58,6 +58,11 @@ impl Compiler {
     }
 }
 
+#[allow(rustc::bad_opt_access)]
+pub fn set_parallel_mode(sopts: &config::UnstableOptions) {
+    rustc_data_structures::sync::set(sopts.threads > 1);
+}
```
Comment (nnethercote) [inline compiler/rustc_interface/src/interface.rs:64]: I can see where this function is called within rustdoc (in `src/librustdoc/lib.rs`), but I can't see where it is called in rustc. Am I overlooking something?

Comment (SparrowLii) [reply compiler/rustc_interface/src/interface.rs:64]: it's in `run_compiler` in `rustc_driver_impl::lib.rs`

Comment (SparrowLii) [reply compiler/rustc_interface/src/interface.rs:64]: In this PR, this part of the modification may be discarded because of rebease

**Comment by SparrowLii** [comment]
Well, it looks like #106930 is still the cause of the blockage. But this time it is not the code in `compiler` but `clippy_utils`

**Comment by SparrowLii** [comment]
This draft will be used to test rustc with `parallel_compierl=true` and with `-Zthreads=1` this month.

**Comment by SparrowLii** [comment]
This draft is far from the current implementation, so I think it can be closed.

**Pull request closed without merging**